File: frontend/src/lib/normalEditing.ts
export type EditingCategory = "blocked_ga" | "synonymous" | "editable";

export interface EditingSite {
  position: number;
  codon: string;
  editedCodon: string;
  aminoAcid: string;
  editedAminoAcid: string;
  category: EditingCategory;
}

export interface InitialArrnaDesign {
  targetPosition: number;
  targetWindow: string;
  arrnaSequence: string;
  windowStart: number;
  windowEnd: number;
  targetIndex: number;
  upstreamLength: number;
  downstreamLength: number;
  hasFullFlanks: boolean;
}

export type AdarEnvironment = "ADAR1" | "ADAR2";
export type NormalArrnaVariantId = "baseline" | "downstream-del10" | "upstream-del28" | "dual-deletion";

export interface NormalArrnaVariant {
  id: NormalArrnaVariantId;
  label: string;
  shortLabel: string;
  description: string;
  arrnaSequence: string;
  alignedGuide: string;
  deletionCoordinates: number[];
}

export interface PairingValidation {
  valid: boolean;
  requiredUpstream: number;
  requiredDownstream: number;
  message: string;
}

const CODON_TABLE: Record<string, string> = {
  TTT: "F", TTC: "F", TTA: "L", TTG: "L",
  TCT: "S", TCC: "S", TCA: "S", TCG: "S",
  TAT: "Y", TAC: "Y", TAA: "*", TAG: "*",
  TGT: "C", TGC: "C", TGA: "*", TGG: "W",
  CTT: "L", CTC: "L", CTA: "L", CTG: "L",
  CCT: "P", CCC: "P", CCA: "P", CCG: "P",
  CAT: "H", CAC: "H", CAA: "Q", CAG: "Q",
  CGT: "R", CGC: "R", CGA: "R", CGG: "R",
  ATT: "I", ATC: "I", ATA: "I", ATG: "M",
  ACT: "T", ACC: "T", ACA: "T", ACG: "T",
  AAT: "N", AAC: "N", AAA: "K", AAG: "K",
  AGT: "S", AGC: "S", AGA: "R", AGG: "R",
  GTT: "V", GTC: "V", GTA: "V", GTG: "V",
  GCT: "A", GCC: "A", GCA: "A", GCG: "A",
  GAT: "D", GAC: "D", GAA: "E", GAG: "E",
  GGT: "G", GGC: "G", GGA: "G", GGG: "G",
};

export function translateCodon(codon: string): string {
  return CODON_TABLE[codon] ?? "?";
}

export function classifyEditingSites(sequence: string): EditingSite[] {
  if (!sequence || sequence.length % 3 !== 0 || /[^ACGT]/.test(sequence)) return [];

  return sequence.split("").flatMap((base, index) => {
    if (base !== "A") return [];
    const codonStart = Math.floor(index / 3) * 3;
    const codon = sequence.slice(codonStart, codonStart + 3);
    const offset = index - codonStart;
    const editedCodon = `${codon.slice(0, offset)}G${codon.slice(offset + 1)}`;
    const aminoAcid = CODON_TABLE[codon];
    const editedAminoAcid = CODON_TABLE[editedCodon];
    const category: EditingCategory = index > 0 && sequence[index - 1] === "G"
      ? "blocked_ga"
      : aminoAcid === editedAminoAcid
        ? "synonymous"
        : "editable";

    return [{
      position: index + 1,
      codon,
      editedCodon,
      aminoAcid,
      editedAminoAcid,
      category,
    }];
  });
}

export function reverseComplement(sequence: string): string {
  const complements: Record<string, string> = { A: "U", C: "G", G: "C", T: "A", U: "A" };
  return sequence
    .split("")
    .reverse()
    .map((base) => complements[base])
    .join("");
}

export function designInitialArrna(
  sequence: string,
  targetPosition: number,
  flankLength = 75,
): InitialArrnaDesign | null {
  const targetIndex = targetPosition - 1;
  if (targetIndex < 0 || targetIndex >= sequence.length || sequence[targetIndex] !== "A") return null;

  const windowStartIndex = Math.max(0, targetIndex - flankLength);
  const windowEndIndex = Math.min(sequence.length, targetIndex + flankLength + 1);
  const targetWindow = sequence.slice(windowStartIndex, windowEndIndex);
  const upstreamLength = targetIndex - windowStartIndex;
  const downstreamLength = windowEndIndex - targetIndex - 1;
  const targetIndexInWindow = targetIndex - windowStartIndex;
  const complementaryGuide = reverseComplement(targetWindow);
  const mismatchIndex = complementaryGuide.length - 1 - targetIndexInWindow;
  const arrnaSequence = `${complementaryGuide.slice(0, mismatchIndex)}C${complementaryGuide.slice(mismatchIndex + 1)}`;

  return {
    targetPosition,
    targetWindow,
    arrnaSequence,
    windowStart: windowStartIndex + 1,
    windowEnd: windowEndIndex,
    targetIndex: targetIndexInWindow,
    upstreamLength,
    downstreamLength,
    hasFullFlanks: upstreamLength === flankLength && downstreamLength === flankLength,
  };
}

export function validateAdarPairing(
  design: InitialArrnaDesign,
  environment: AdarEnvironment,
): PairingValidation {
  const requiredDownstream = environment === "ADAR1" ? 4 : 6;
  const requiredUpstream = environment === "ADAR1" ? 20 : 11;
  const problems: string[] = [];
  if (design.downstreamLength < requiredDownstream) {
    problems.push(`arRNA 5′ side requires ${requiredDownstream} paired nt, but only ${design.downstreamLength} nt are available`);
  }
  if (design.upstreamLength < requiredUpstream) {
    problems.push(`arRNA 3′ side requires ${requiredUpstream} paired nt, but only ${design.upstreamLength} nt are available`);
  }
  return {
    valid: problems.length === 0,
    requiredUpstream,
    requiredDownstream,
    message: problems.join("; "),
  };
}

function createDeletionVariant(
  design: InitialArrnaDesign,
  id: NormalArrnaVariantId,
  label: string,
  shortLabel: string,
  description: string,
  deletionCoordinates: number[],
): NormalArrnaVariant {
  const deleted = new Set(deletionCoordinates);
  const alignedGuide = design.arrnaSequence.split("").reverse().map((base, targetIndex) => {
    const relativeCoordinate = targetIndex - design.targetIndex;
    return deleted.has(relativeCoordinate) ? "-" : base;
  }).join("");
  const arrnaSequence = design.arrnaSequence.split("").filter((_, arrnaIndex) => {
    const targetIndex = design.arrnaSequence.length - 1 - arrnaIndex;
    const relativeCoordinate = targetIndex - design.targetIndex;
    return !deleted.has(relativeCoordinate);
  }).join("");
  return { id, label, shortLabel, description, arrnaSequence, alignedGuide, deletionCoordinates };
}

export function designNormalArrnaVariants(design: InitialArrnaDesign): NormalArrnaVariant[] {
  const variants: NormalArrnaVariant[] = [createDeletionVariant(
    design,
    "baseline",
    "Baseline arRNA",
    "No deletion",
    "A–C target mismatch; all other positions are complementary.",
    [],
  )];
  const downstreamDeletion = Array.from({ length: 10 }, (_, index) => 34 + index);
  const upstreamDeletion = Array.from({ length: 28 }, (_, index) => -31 - index);

  if (design.downstreamLength >= 75) {
    variants.push(createDeletionVariant(
      design,
      "downstream-del10",
      "Downstream +34 deletion arRNA",
      "+34 deletion 10",
      "Deletes the 10 arRNA bases aligned to A+34–A+43; 32 distal paired nt remain.",
      downstreamDeletion,
    ));
  }
  if (design.upstreamLength >= 75) {
    variants.push(createDeletionVariant(
      design,
      "upstream-del28",
      "Upstream −31 deletion arRNA",
      "−31 deletion 28",
      "Deletes the 28 arRNA bases aligned to A−31–A−58; 17 distal paired nt remain.",
      upstreamDeletion,
    ));
  }
  if (design.downstreamLength >= 75 && design.upstreamLength >= 75) {
    variants.push(createDeletionVariant(
      design,
      "dual-deletion",
      "Dual-deletion arRNA",
      "Dual deletion",
      "Combines A+34 deletion 10 and A−31 deletion 28.",
      [...downstreamDeletion, ...upstreamDeletion],
    ));
  }
  return variants;
}

File: backend/app/modules/normal_editing.py
from app.schemas.design import (
    DesignProvenance,
    EditingSiteResult,
    EngineInfo,
    NormalEditingCandidate,
    NormalEditingRequest,
    NormalEditingResult,
    NormalizedSequenceInput,
    PairingRequirements,
    RuleSetInfo,
)

CODON_TABLE: dict[str, str] = {
    "TTT": "F", "TTC": "F", "TTA": "L", "TTG": "L",
    "TCT": "S", "TCC": "S", "TCA": "S", "TCG": "S",
    "TAT": "Y", "TAC": "Y", "TAA": "*", "TAG": "*",
    "TGT": "C", "TGC": "C", "TGA": "*", "TGG": "W",
    "CTT": "L", "CTC": "L", "CTA": "L", "CTG": "L",
    "CCT": "P", "CCC": "P", "CCA": "P", "CCG": "P",
    "CAT": "H", "CAC": "H", "CAA": "Q", "CAG": "Q",
    "CGT": "R", "CGC": "R", "CGA": "R", "CGG": "R",
    "ATT": "I", "ATC": "I", "ATA": "I", "ATG": "M",
    "ACT": "T", "ACC": "T", "ACA": "T", "ACG": "T",
    "AAT": "N", "AAC": "N", "AAA": "K", "AAG": "K",
    "AGT": "S", "AGC": "S", "AGA": "R", "AGG": "R",
    "GTT": "V", "GTC": "V", "GTA": "V", "GTG": "V",
    "GCT": "A", "GCC": "A", "GCA": "A", "GCG": "A",
    "GAT": "D", "GAC": "D", "GAA": "E", "GAG": "E",
    "GGT": "G", "GGC": "G", "GGA": "G", "GGG": "G",
}

ENGINE = EngineInfo(name="leaper-arrna-engine", version="0.2.0")
NORMAL_RULESET = RuleSetInfo(
    id="leaper-normal-editing",
    version="0.1.0",
)


def reverse_complement_rna(sequence: str) -> str:
    return sequence.translate(str.maketrans("ACGT", "UGCA"))[::-1]


def _editing_site(sequence: str, target_position: int) -> EditingSiteResult:
    target_index = target_position - 1
    codon_start = (target_index // 3) * 3
    codon = sequence[codon_start:codon_start + 3]
    offset = target_index - codon_start
    edited_codon = f"{codon[:offset]}G{codon[offset + 1:]}"
    amino_acid = CODON_TABLE[codon]
    edited_amino_acid = CODON_TABLE[edited_codon]
    if target_index > 0 and sequence[target_index - 1] == "G":
        category = "blocked_ga"
    elif amino_acid == edited_amino_acid:
        category = "synonymous"
    else:
        category = "editable"
    return EditingSiteResult(
        position=target_position,
        codon=codon,
        edited_codon=edited_codon,
        amino_acid=amino_acid,
        edited_amino_acid=edited_amino_acid,
        category=category,
    )


def _candidate(
    arrna_sequence: str,
    target_index: int,
    candidate_id: str,
    label: str,
    short_label: str,
    description: str,
    deletion_coordinates: list[int],
) -> NormalEditingCandidate:
    deleted = set(deletion_coordinates)
    aligned_guide = "".join(
        "-" if target_index_in_window - target_index in deleted else base
        for target_index_in_window, base in enumerate(reversed(arrna_sequence))
    )
    sequence = "".join(
        base
        for arrna_index, base in enumerate(arrna_sequence)
        if len(arrna_sequence) - 1 - arrna_index - target_index not in deleted
    )
    return NormalEditingCandidate(
        id=candidate_id,
        label=label,
        short_label=short_label,
        description=description,
        arrna_sequence=sequence,
        aligned_guide=aligned_guide,
        deletion_coordinates=deletion_coordinates,
    )


def design_normal_editing(
    payload: NormalEditingRequest,
    flank_length: int = 75,
) -> NormalEditingResult:
    sequence = payload.sequence
    target_index = payload.target_position - 1
    start_index = max(0, target_index - flank_length)
    end_index = min(len(sequence), target_index + flank_length + 1)
    target_window = sequence[start_index:end_index]
    target_index_in_window = target_index - start_index
    upstream_length = target_index_in_window
    downstream_length = len(target_window) - target_index_in_window - 1

    complementary_guide = reverse_complement_rna(target_window)
    mismatch_index = len(complementary_guide) - 1 - target_index_in_window
    baseline_arrna = (
        f"{complementary_guide[:mismatch_index]}C"
        f"{complementary_guide[mismatch_index + 1:]}"
    )

    candidates = [_candidate(
        baseline_arrna,
        target_index_in_window,
        "baseline",
        "Baseline arRNA",
        "No deletion",
        "A-C target mismatch with otherwise complete complementary pairing.",
        [],
    )]
    downstream_deletion = list(range(34, 44))
    upstream_deletion = list(range(-31, -59, -1))
    if downstream_length >= flank_length:
        candidates.append(_candidate(
            baseline_arrna,
            target_index_in_window,
            "downstream-del10",
            "Downstream +34 deletion arRNA",
            "+34 deletion 10",
            "Deletes arRNA bases aligned to target coordinates A+34 through A+43.",
            downstream_deletion,
        ))
    if upstream_length >= flank_length:
        candidates.append(_candidate(
            baseline_arrna,
            target_index_in_window,
            "upstream-del28",
            "Upstream -31 deletion arRNA",
            "-31 deletion 28",
            "Deletes arRNA bases aligned to target coordinates A-31 through A-58.",
            upstream_deletion,
        ))
    if downstream_length >= flank_length and upstream_length >= flank_length:
        candidates.append(_candidate(
            baseline_arrna,
            target_index_in_window,
            "dual-deletion",
            "Dual-deletion arRNA",
            "Dual deletion",
            "Combines the A+34 deletion 10 and A-31 deletion 28 gaps.",
            [*downstream_deletion, *upstream_deletion],
        ))

    required_downstream = 4 if payload.adar_environment == "ADAR1" else 6
    required_upstream = 20 if payload.adar_environment == "ADAR1" else 11
    problems: list[str] = []
    if downstream_length < required_downstream:
        problems.append(
            f"arRNA 5' side requires {required_downstream} paired nt; "
            f"only {downstream_length} are available"
        )
    if upstream_length < required_upstream:
        problems.append(
            f"arRNA 3' side requires {required_upstream} paired nt; "
            f"only {upstream_length} are available"
        )
    warnings = [] if upstream_length == flank_length and downstream_length == flank_length else [
        "The target does not have the complete 75-nt context on both sides."
    ]
    if problems:
        warnings.append("; ".join(problems))

    return NormalEditingResult(
        normalized_input=NormalizedSequenceInput(
            sequence=sequence,
            length=len(sequence),
            target_position=payload.target_position,
        ),
        reference=payload.reference,
        editing_site=_editing_site(sequence, payload.target_position),
        target_window=target_window,
        window_start=start_index + 1,
        window_end=end_index,
        target_index=target_index_in_window,
        upstream_length=upstream_length,
        downstream_length=downstream_length,
        has_full_flanks=upstream_length == flank_length and downstream_length == flank_length,
        pairing=PairingRequirements(
            adar_environment=payload.adar_environment,
            required_upstream=required_upstream,
            required_downstream=required_downstream,
            is_satisfied=not problems,
            message="; ".join(problems) if problems else "Pairing requirements are satisfied.",
        ),
        candidates=candidates,
        warnings=warnings,
        provenance=DesignProvenance(engine=ENGINE, ruleset=NORMAL_RULESET),
    )
